File: backend/workflows/ingest_graph.py
import logging

from langgraph.config import get_stream_writer
from langgraph.graph import END, StateGraph
from qdrant_client import models

from clients.qdrant import qdrant_client
from ingestion.chunking import chunk_transcript as split_transcript_into_chunks
from ingestion.embedder import generate_embeddings
from utils.youtube import get_transcript
from workflows.states import IngestState

logger = logging.getLogger(__name__)


def _emit_progress(update: dict) -> None:
    writer = get_stream_writer()
    if writer:
        writer(update)


def extract_transcript(state: IngestState) -> IngestState:
    progress = {"status": "Extracting Transcript", "total_chunks": 0, "processed_chunks": 0}
    _emit_progress(progress)

    try:
        transcript = get_transcript(state["video_id"])
        return {**progress, "transcript": transcript, "total_chunks": len(transcript)}
    except Exception as exc:
        logger.exception("Failed to extract transcript for %s", state["video_id"])
        error_msg = (
            "Could not retrieve transcript. The video might not have captions or is restricted."
            if "transcript" in str(exc).lower()
            else "Failed to process the video. Please ensure it has captions and try again."
        )
        return {
            "status": "failed",
            "error": error_msg,
            "completed": True,
        }


def chunk_transcript(state: IngestState) -> IngestState:
    progress = {"status": "Chunking Transcript", "total_chunks": state["total_chunks"], "processed_chunks": 0}
    _emit_progress(progress)

    chunks = split_transcript_into_chunks(state["transcript"])
    return {
        **progress,
        "chunks": chunks,
        "total_chunks": len(chunks),
    }


def embed_and_store(state: IngestState, batch_size: int = 100) -> IngestState:
    chunks = state["chunks"]
    total_chunks = len(chunks)
    collection_name = f"video_{state['video_id'].replace('-', '_')}"

    _emit_progress(
        {
            "status": "Processing Chunks",
            "total_chunks": total_chunks,
            "processed_chunks": 0,
        }
    )

    try:
        if not qdrant_client.collection_exists(collection_name=collection_name):
            qdrant_client.create_collection(
                collection_name=collection_name,
                vectors_config=models.VectorParams(size=3072, distance=models.Distance.COSINE),
            )

        existing_count = qdrant_client.count(collection_name=collection_name).count
        if existing_count >= total_chunks:
            _emit_progress(
                {
                    "status": "Completed Chunks Processing",
                    "total_chunks": total_chunks,
                    "processed_chunks": existing_count,
                }
            )
            return {
                "status": "Completed Chunks Processing",
                "total_chunks": total_chunks,
                "processed_chunks": existing_count,
                "completed": True,
            }

        for i in range(existing_count, total_chunks, batch_size):
            batch_chunks = chunks[i : i + batch_size]
            texts = [chunk["text"] for chunk in batch_chunks]
            embeddings = generate_embeddings(texts)
            points = [
                models.PointStruct(
                    id=i + j,
                    vector=embedding,
                    payload={"text": batch_chunks[j]["text"]},
                )
                for j, embedding in enumerate(embeddings)
            ]

            qdrant_client.upload_points(
                collection_name=collection_name,
                points=points,
                wait=True,
            )

            processed = i + len(batch_chunks)
            _emit_progress(
                {
                    "status": "Processing...",
                    "total_chunks": total_chunks,
                    "processed_chunks": processed,
                }
            )

        _emit_progress(
            {
                "status": "Completed Processing",
                "total_chunks": total_chunks,
                "processed_chunks": total_chunks,
            }
        )
        return {
            "status": "Completed Processing",
            "total_chunks": total_chunks,
            "processed_chunks": total_chunks,
            "completed": True,
        }
    except Exception:
        logger.exception("Failed to embed/store chunks for %s", state["video_id"])
        error_msg = "An error occurred during embedding processing."
        return {
            "status": "Failed Processing",
            "total_chunks": total_chunks,
            "error": error_msg,
            "completed": True,
        }


def route_after_extract(state: IngestState) -> str:
    if state.get("status") == "failed":
        return "end"
    return "chunk_transcript"


def build_ingest_graph():
    graph = StateGraph(IngestState)
    graph.add_node("extract_transcript", extract_transcript)
    graph.add_node("chunk_transcript", chunk_transcript)
    graph.add_node("embed_and_store", embed_and_store)

    graph.set_entry_point("extract_transcript")
    graph.add_conditional_edges(
        "extract_transcript",
        route_after_extract,
        {"end": END, "chunk_transcript": "chunk_transcript"},
    )
    graph.add_edge("chunk_transcript", "embed_and_store")
    graph.add_edge("embed_and_store", END)

    return graph.compile()

File: frontend/src/lib/rag/workflows/ingest-graph.ts
import { END, START, StateGraph } from "@langchain/langgraph";
import type { LangGraphRunnableConfig } from "@langchain/langgraph";
import { EMBEDDING_DIMENSIONS, EMBEDDING_MODEL, QDRANT_UPSERT_BATCH_SIZE } from "../constants";
import {
  collectionNameForVideo,
  getQdrantClient,
  markCollectionReady,
} from "../clients/qdrant";
import { chunkTranscript } from "../ingestion/chunking";
import { generateEmbeddings } from "../ingestion/embedder";
import { getTranscript } from "../ingestion/transcript";
import type { IngestStreamEvent, TextChunk } from "../types";
import { formatTimestampRange } from "../timestamp";
import { IngestStateAnnotation, type IngestState } from "./states";

function emit(config: LangGraphRunnableConfig, event: IngestStreamEvent) {
  config.writer?.(event);
}

function progress(
  config: LangGraphRunnableConfig,
  update: {
    status: string;
    total_chunks: number;
    processed_chunks: number;
  },
) {
  emit(config, { type: "progress", ...update });
}

function embedTextForChunk(chunk: TextChunk): string {
  const range = formatTimestampRange(chunk.startSeconds, chunk.endSeconds);
  return `Video transcript segment [${range}]: ${chunk.text}`;
}

async function collectionNeedsFullReembed(
  collectionName: string,
  totalChunks: number,
): Promise<boolean> {
  const qdrant = getQdrantClient();
  const exists = await qdrant.collectionExists(collectionName);
  if (!exists.exists) {
    return true;
  }

  const info = await qdrant.getCollection(collectionName);
  const vectors = info.config?.params?.vectors;
  const currentSize =
    vectors && typeof vectors === "object" && "size" in vectors
      ? vectors.size
      : undefined;

  if (currentSize !== EMBEDDING_DIMENSIONS) {
    return true;
  }

  const scroll = await qdrant.scroll(collectionName, {
    limit: 1,
    with_payload: true,
  });
  const point = scroll.points[0];
  if (!point?.payload) {
    return true;
  }

  const payload = point.payload as {
    embedding_model?: string;
    total_chunks?: number;
  };

  return (
    payload.embedding_model !== EMBEDDING_MODEL ||
    payload.total_chunks !== totalChunks
  );
}

async function extractTranscriptNode(
  state: IngestState,
  config: LangGraphRunnableConfig,
) {
  progress(config, {
    status: "Extracting Transcript",
    total_chunks: 0,
    processed_chunks: 0,
  });

  try {
    const transcript = await getTranscript(state.videoId);
    return {
      status: "Extracting Transcript",
      transcript,
      totalChunks: transcript.length,
      processedChunks: 0,
    };
  } catch (error) {
    console.error("Failed to extract transcript:", error);
    const message =
      error instanceof Error && error.message.toLowerCase().includes("transcript")
        ? "Could not retrieve transcript. The video might not have captions or is restricted."
        : "Failed to process the video. Please ensure it has captions and try again.";

    emit(config, {
      type: "error",
      video_id: state.videoId,
      status: "failed",
      error: message,
      completed: true,
    });

    return {
      status: "failed",
      error: message,
      completed: true,
    };
  }
}

function routeAfterExtract(state: IngestState) {
  if (state.status === "failed") {
    return END;
  }
  return "chunk_transcript";
}

async function chunkTranscriptNode(
  state: IngestState,
  config: LangGraphRunnableConfig,
) {
  progress(config, {
    status: "Chunking Transcript",
    total_chunks: state.totalChunks,
    processed_chunks: 0,
  });

  const chunks = chunkTranscript(state.transcript);
  return {
    status: "Chunking Transcript",
    chunks,
    totalChunks: chunks.length,
    processedChunks: 0,
  };
}

async function embedAndStoreNode(
  state: IngestState,
  config: LangGraphRunnableConfig,
) {
  const chunks = state.chunks;
  const totalChunks = chunks.length;
  const collectionName = collectionNameForVideo(state.videoId);
  const qdrant = getQdrantClient();

  progress(config, {
    status: "Processing Chunks",
    total_chunks: totalChunks,
    processed_chunks: 0,
  });

  try {
    const needsFullReembed = await collectionNeedsFullReembed(collectionName, totalChunks);
    let existingCount = 0;

    if (needsFullReembed) {
      const exists = await qdrant.collectionExists(collectionName);
      if (exists.exists) {
        await qdrant.deleteCollection(collectionName);
      }
      await qdrant.createCollection(collectionName, {
        vectors: {
          size: EMBEDDING_DIMENSIONS,
          distance: "Cosine",
        },
      });
    } else {
      const countResult = await qdrant.count(collectionName, { exact: true });
      existingCount = countResult.count ?? 0;
    }

    if (existingCount >= totalChunks) {
      markCollectionReady(state.videoId);
      progress(config, {
        status: "Completed Chunks Processing",
        total_chunks: totalChunks,
        processed_chunks: existingCount,
      });
      emit(config, {
        type: "complete",
        video_id: state.videoId,
        status: "Completed Chunks Processing",
        total_chunks: totalChunks,
        processed_chunks: existingCount,
        completed: true,
      });
      return {
        status: "Completed Chunks Processing",
        totalChunks,
        processedChunks: existingCount,
        completed: true,
      };
    }

    const batchSize = QDRANT_UPSERT_BATCH_SIZE;
    for (let i = existingCount; i < totalChunks; i += batchSize) {
      const batchChunks = chunks.slice(i, i + batchSize);
      const texts = batchChunks.map(embedTextForChunk);
      const embeddings = await generateEmbeddings(texts);
      const isLastBatch = i + batchChunks.length >= totalChunks;

      await qdrant.upsert(collectionName, {
        wait: isLastBatch,
        points: embeddings.map((vector, index) => ({
          id: i + index,
          vector,
          payload: {
            text: batchChunks[index].text,
            context_text: batchChunks[index].contextText,
            start_seconds: batchChunks[index].startSeconds,
            end_seconds: batchChunks[index].endSeconds,
            chunk_index: batchChunks[index].chunkIndex,
            embedding_model: EMBEDDING_MODEL,
            total_chunks: totalChunks,
          },
        })),
      });

      const processed = i + batchChunks.length;
      progress(config, {
        status: "Processing...",
        total_chunks: totalChunks,
        processed_chunks: processed,
      });
    }

    markCollectionReady(state.videoId);

    progress(config, {
      status: "Completed Processing",
      total_chunks: totalChunks,
      processed_chunks: totalChunks,
    });

    emit(config, {
      type: "complete",
      video_id: state.videoId,
      status: "Completed Processing",
      total_chunks: totalChunks,
      processed_chunks: totalChunks,
      completed: true,
    });

    return {
      status: "Completed Processing",
      totalChunks,
      processedChunks: totalChunks,
      completed: true,
    };
  } catch (error) {
    console.error("Failed to embed/store chunks:", error);
    const message =
      error instanceof Error && error.message.toLowerCase().includes("timeout")
        ? "Vector database timed out. Please try again."
        : "An error occurred during embedding processing.";

    emit(config, {
      type: "error",
      video_id: state.videoId,
      status: "Failed Processing",
      total_chunks: totalChunks,
      error: message,
      completed: true,
    });

    return {
      status: "Failed Processing",
      totalChunks,
      error: message,
      completed: true,
    };
  }
}

export function buildIngestGraph() {
  const graph = new StateGraph(IngestStateAnnotation)
    .addNode("extract_transcript", extractTranscriptNode)
    .addNode("chunk_transcript", chunkTranscriptNode)
    .addNode("embed_and_store", embedAndStoreNode)
    .addEdge(START, "extract_transcript")
    .addConditionalEdges("extract_transcript", routeAfterExtract)
    .addEdge("chunk_transcript", "embed_and_store")
    .addEdge("embed_and_store", END);

  return graph.compile();
}
